File: Blind150/7.Tree/Python/level_order_traversal.py
# https://leetcode.com/problems/binary-tree-level-order-traversal/description/


from typing import Optional, List
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:

        if not root:
            return []

        d = deque([root])

        ans = []

        while d:

            level_list = []

            for _ in range(len(d)):
                n = d.popleft()
                level_list.append(n.val)

                if n.left:
                    d.append(n.left)

                if n.right:
                    d.append(n.right)

            ans.append(level_list)

        return ans

File: Blind150/7.Tree/Go/level_order_traversal.go
// https://leetcode.com/problems/binary-tree-level-order-traversal/description/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrder(root *TreeNode) [][]int {

	if root == nil {
		return [][]int{}
	}

	dq := []*TreeNode{root}
	result := [][]int{}

	for len(dq) > 0 {

		levelList := []int{}
		levelSize := len(dq)

		for i := 0; i < levelSize; i++ {

			curNode := dq[0] // take the first from left from the queue

			dq = dq[1:] // remove the first element from the queue

			levelList = append(levelList, curNode.Val)

			if curNode.Left != nil {

				dq = append(dq, curNode.Left)
			}

			if curNode.Right != nil {

				dq = append(dq, curNode.Right)
			}

		}

		result = append(result, levelList)

	}

	return result

}
